pkg/cmd/tests/tests_run.go

```diff
@@ -164,38 +258,198 @@ func (o *RunOptions) run(ctx context.Context, streams genericclioptions.IOStream
-	gconfig.GinkgoConfig.SkipMeasurements = o.SkipMeasurements
-	gconfig.GinkgoConfig.RandomSeed = o.RandomSeed
-	gconfig.GinkgoConfig.DryRun = o.DryRun
-	gconfig.DefaultReporterConfig.Verbose = !o.Quiet
-	gconfig.DefaultReporterConfig.NoColor = !o.Color
 
 	// Not configurable. We are opinionated about these.
 
 	// Prevents growing a dependency.
-	gconfig.GinkgoConfig.RandomizeAllSpecs = true
+	suiteConfig.RandomizeAllSpecs = true
 	// Better context and it's required for nested assertions. Offset doesn't really solve it as it omits the nested function line.
-	gconfig.DefaultReporterConfig.FullTrace = true
-	gconfig.DefaultReporterConfig.SlowSpecThreshold = 60 // seconds
+	reporterConfig.FullTrace = true
+	reporterConfig.SlowSpecThreshold = 10 * time.Minute
 
 	gomega.RegisterFailHandler(ginkgo.Fail)
 
-	var r []ginkgo.Reporter
 	if len(o.ArtifactsDir) > 0 {
-		r = append(r, greporters.NewJUnitReporter(path.Join(o.ArtifactsDir, fmt.Sprintf("e2e_%02d.xml", gconfig.GinkgoConfig.ParallelNode))))
+		reporterConfig.JUnitReport = path.Join(o.ArtifactsDir, "e2e.junit.xml")
+		reporterConfig.JSONReport = path.Join(o.ArtifactsDir, "e2e.json")
+	}
+
+	suiteConfig.ParallelTotal = o.Parallelism
+	if suiteConfig.ParallelTotal == 0 {
+		if o.DryRun {
+			suiteConfig.ParallelTotal = 1
+		} else {
+			suiteConfig.ParallelTotal = o.SelectedSuite.DefaultParallelism
+		}
+	}
+
+	if suiteConfig.ParallelTotal <= 1 || o.ParallelShard != 0 {
+		if suiteConfig.ParallelTotal > 1 {
+			suiteConfig.ParallelHost = o.ParallelServerAddress
+			suiteConfig.ParallelProcess = o.ParallelShard
+		}
+
+		klog.InfoS("Running specs")
+		passed := ginkgo.RunSpecs(&fakeT{}, suite, suiteConfig, reporterConfig)
+		if !passed {
+			return fmt.Errorf("test suite %q failed", suite)
+		}
+
+		return nil
+	}
+
+	server, err := parallel_support.NewServer(suiteConfig.ParallelTotal, reporters.NewDefaultReporter(reporterConfig, formatter.ColorableStdOut))
+	if err != nil {
+		return fmt.Errorf("can't create parallel spec server")
+	}
+	server.Start()
+	defer server.Close()
+
+	commonArgs := make([]string, 0, 2+len(os.Args))
+	if len(os.Args) > 1 {
+		commonArgs = append(commonArgs, os.Args[1:]...)
+	}
+	commonArgs = append(commonArgs, fmt.Sprintf("--%s=%v", parallelServerAddressFlagKey, server.Address()))
+	commonArgs = append(commonArgs, fmt.Sprintf("--%s=%v", cmdutil.FlagLogLevelKey, o.ParallelLogLevel))
+
+	type cmdEntry struct {
+		id  int
+		cmd *exec.Cmd
+		out *bytes.Buffer
+	}
+	cmdEntries := make([]*cmdEntry, 0, suiteConfig.ParallelTotal)
+	for i := 1; i <= suiteConfig.ParallelTotal; i++ {
+		suiteConfig := suiteConfig
+		suiteConfig.ParallelProcess = i
+
+		args := make([]string, 0, len(commonArgs)+1)
+		args = append(args, commonArgs...)
+		args = append(args, fmt.Sprintf("--%s=%v", parallelShardFlagKey, i))
+
+		buf := &bytes.Buffer{}
+		cmd := exec.CommandContext(ctx, os.Args[0], args...)
+		cmd.Stdout = buf
+		cmd.Stderr = buf
+		cmdEntries = append(cmdEntries, &cmdEntry{
```

Comment:
what is this printing out? I thought it's test progress, but it isn't printed anywhere on happy path.

Reply:
it's printed at the end only when the subprocess fails

Reply:
I meant stdout, but looks like test progress is printed by the server